### PR TITLE
Logan/professor question

### PR DESCRIPTION
**Pull request by loganallen**
Apologies for such a HUGE PR, future ones will be small in the future. This PR establishes groundwork for the professor-side interface. I wanted to make an initial PR to get this bulk code out, then fine-tune it with more network requests. This PR contains the following:

- `LectureDashboard` (TODO: connect with redux and server to load lectures)
- `LecturePage`
    - Appears when clicking `Create Lecture` or editing an existing one (TODO) in LectureDashboard
    - Displays list of questions for the lecture (TODO: fetch list from server)
- `LectureQuestionCreator`
    - Empty after clicking `Create Question`, populated with data after clicking an existing question
    - Toggles between 4 QuestionTypes (MC fully implemented with saving to DB)

Notes: Uses Redux to handle network requests (so far just `save_question`) and a global store to be used for all courses, lectures, and questions/lecture for a given professor. Some components (like `LectureQuestionCreator` take data in as props and uses local state when editing, updating the global store only on `save`

**Uses `semantic-ui-react` for styling with additional per-component styles. Temporarily included the stylesheet link in `index.html` because CSS loaders were blocking development.

image
image

### CODE REVIEW

**Comment by Jma353** [comment]
Yoooo this is hugeeeee - let's shoot for smaller in the future.  Will review later tonight

**Comment by Jma353** [comment]
@meganle can you start looking at this for JS style, as well as to start familiarizing yourself with some of the stuff that happens on frontend.  Would be very dope.

File: src/components/common/constants.js [hunk 1]
```diff
@@ -1,10 +1,34 @@
...
+export type QuestionType = 'MULTIPLE_CHOICE' | 'FREE_RESPONSE' |
+  'MULTIPLE_ANSWER' | 'RANKING'
+
+const QUESTION_TYPES = {
+  MULTIPLE_CHOICE: 'MULTIPLE_CHOICE',
+  FREE_RESPONSE: 'FREE_RESPONSE',
+  MULTIPLE_ANSWER: 'MULTIPLE_ANSWER',
+  RANKING: 'RANKING'
+};
+// const QUESTION_TYPES = keyMirror({
```
Comment (Jma353) [inline src/components/common/constants.js]: Remove

File: src/components/lecture/EditLectureModal.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import React from 'react';
+import {
+  Modal,
+  Input,
+  Button
+} from 'semantic-ui-react';
+
+type Props = {
+  open: boolean,
+  onClose: () => void,
```
Comment (Jma353) [inline src/components/lecture/EditLectureModal.js]: Functions below non-functional props

File: src/components/lecture/EditLectureModal.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  onSave: (Object) => void,
+  lectureTitle: string
+};
```
Comment (Jma353) [inline src/components/lecture/EditLectureModal.js]: Add newline

File: src/components/lecture/EditLectureModal.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  state: State;
+
+  constructor (props: Props) {
+    super(props);
+    this.state = {
+      lectureTitle: props.lectureTitle
+    };
+  }
+
+  onLectureTitleChange = (event: Object): void => {
```
Comment (Jma353) [inline src/components/lecture/EditLectureModal.js]: Don't use `Object` for `event`, see [this](https://flow.org/en/docs/react/events/)

File: src/components/lecture/LecturePage.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  Card
+} from 'semantic-ui-react';
+import actions from '../actions';
+import type { QuestionType } from '../common/constants';
+
+import ClickerPage from '../common/ClickerPage';
+import EditLectureModal from './EditLectureModal';
+import LectureQuestionCreator from './LectureQuestionCreator';
+
+type Props = {
```
Comment (Jma353) [inline src/components/lecture/LecturePage.js]: Without even looking at the code for this, there seems to be a lot of props.  This is sometimes fine if most of them are optional and the component is a robust one that can be reused in a diverse array of situations, but this is a highly specialized component.  Maybe some of these functions passed down could be one function with a parameter indicating the operation?  Maybe some of these details can be a typed JSON indicating the data being passed down?  These techniques eliminate prop-overload.

File: src/components/lecture/LecturePage.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+      <Button.Group basic compact vertical icon labeled>
+        <Button
+          onClick={ () => this.props.onToggleEditLectureModal(true) }
+          compact
+          content='Edit'
+          icon='edit'
+        />
+        <EditLectureModal
+          open={this.props.editLectureModalOpen}
+          onClose={ () => this.props.onToggleEditLectureModal(false) }
```
Comment (Jma353) [inline src/components/lecture/LecturePage.js]: Should be defined as a function in the component: 

```javascript
handleCloseModal = (): void => {
  this.props.onToggleEditLectureModal(false);
}
```

File: src/components/lecture/LecturePage.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+      <Card.Group>
+        {questionCards}
+      </Card.Group>
+    );
+  }
+
+  _sideBar = (): Segment => (
+    <Segment raised>
+      <Button
+        onClick={ () => this.props.onEditQuestion(null) }
```
Comment (Jma353) [inline src/components/lecture/LecturePage.js]:
```javascript
handleClickSideBar = (): void => {
  this.props.onEditQuestion(null);
}
```

File: src/components/lecture/LecturePage.js [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+// Map store to props
+const stateProps = (store: Object) => {
+  return {
+    ...store.lecture
+  };
+};
+
+// Map dispatch actions to props
```
Comment (Jma353) [inline src/components/lecture/LecturePage.js]: This, this right here, is the one thing I hate about `Redux`.... Everything else I like.  @mrkev any advice on how to reduce the violations of `DRY` that this imposes?

File: src/components/lecture/LectureQuestionCreator.js [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+class LectureQuestionCreator extends React.Component<void, Props, State> {
+  props: Props;
+  state: State;
+
+  constructor (props: Props) {
+    super(props);
+    var state = {};
+    if (props.questionData) {
+      state = {
```
Comment (Jma353) [inline src/components/lecture/LectureQuestionCreator.js]: `this.state`

File: src/components/lecture/LectureQuestionCreator.js [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+        questionId: props.questionData.id,
+        questionText: props.questionData.text,
+        questionType: props.questionData.type,
+        activeQuestionTypeIndex: this.mapQuestionTypeToMenuIndex(props.questionData.type)
+      }
+    } else {
```
Comment (Jma353) [inline src/components/lecture/LectureQuestionCreator.js]: Will this ever be called?  `questionData` isn't optional

File: src/components/lecture/LectureQuestionCreator.js [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+      state = {
```
Comment (Jma353) [inline src/components/lecture/LectureQuestionCreator.js]: `this.state`

File: src/components/lecture/LectureQuestionCreator.js [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+  questionData: Object
+};
+
+class LectureQuestionCreator extends React.Component<void, Props, State> {
+  props: Props;
+  state: State;
+
+  constructor (props: Props) {
+    super(props);
+    var state = {};
```
Comment (Jma353) [inline src/components/lecture/LectureQuestionCreator.js]: Unnecessary variable.

File: src/components/lecture/LectureQuestionCreator.js [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+    } else {
+      state = {
+        questionText: '',
+        questionType: QUESTION_TYPES.MULTIPLE_CHOICE,
+        activeQuestionTypeIndex: 0
+      }
+    }
+    this.state = {
+      ...state,
+      questionData: this.initializeQuestionData(props.questionData)
```
Comment (Jma353) [inline src/components/lecture/LectureQuestionCreator.js]: You add `questionData` regardless of which `if-else` branch you take, so no need to use spread to add it here.

File: src/components/lecture/LectureQuestionCreator.js [hunk 14]
```diff
@@ -0,0 +1,321 @@
+//@flow
```
Comment (Jma353) [inline src/components/lecture/LectureQuestionCreator.js]: Something I notice globally about this file: 
don't use `var`.  `let` is the new standard.

File: src/components/lecture/LectureQuestionCreator.js [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+  initializeQuestionData (data?: Object): Object {
+    const questionData = {
+      [QUESTION_TYPES.MULTIPLE_CHOICE]: { options: [], answer: '' },
+      [QUESTION_TYPES.FREE_RESPONSE]: { optionalAnswer: '' },
+      [QUESTION_TYPES.MULTIPLE_ANSWER]: { options: [], answer: [] },
+      [QUESTION_TYPES.RANKING]: { options: [], answer: [] }
+    };
+    if (data) {
+      var updatedData = Object.assign({}, data);
+      delete updatedData.id;
```
Comment (Jma353) [inline src/components/lecture/LectureQuestionCreator.js]: @mrkev is `delete` good?  I always disapproved of it as it seems to encourage mutability.

File: src/components/lecture/LectureQuestionCreator.js (hunk 10 above)
Comment (loganallen) [reply src/components/lecture/LectureQuestionCreator.js]: It is `null` when creating a new question

File: src/components/lecture/LecturePage.js (hunk 5 above)
Comment (loganallen) [reply src/components/lecture/LecturePage.js]: How about grouping prop functions into `actionHandlers` objects. So in `mapDispatchToProps` I return two `handlers` objects which contain the functions relating to that type of component `const editQuestionHandlers = {
    onEditQuestion,
    onCancelEditQuestion,
    onSaveQuestion,
    onDeleteQuestion
  };`

File: src/components/lecture/LecturePage.js [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+const dispatchProps = (dispatch: Function) => {
```
Comment (mrkev) [inline src/components/lecture/LecturePage.js]: `import type { Dispatch as ReduxDispatch } from 'redux';`

~Function~ ReduxDispatch<Action>

Ideally we'd have a type Action that unites all possible actions. This is a massive PR to refactor though, so I'm fine with having this as a TODO/Issue to hasten the merging and reduce possible conflicts.

File: src/components/lecture/LecturePage.js (hunk 8 above)
Comment (mrkev) [reply src/components/lecture/LecturePage.js]: I mean, we could always call dispatch directly without passing the action in as props; this is just one way to architect things.

File: src/components/actions/LectureActions.js [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    type: LectureActionTypes.SAVE_LECTURE,
+    data: data
+  };
+};
+
+const toggleQuestionModal = (show: boolean) => {
+  return {
+    type: LectureActionTypes.TOGGLE_QUESTION_MODAL,
+    show: show
+  }
```
Comment (meganle) [inline src/components/actions/LectureActions.js]: semicolons

File: src/components/actions/LectureActions.js [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    questions: questions
+  };
+};
+
+const saveQuestionFailure = (error: Object) => {
+  return {
+    type: LectureActionTypes.SAVE_QUESTION,
+    status: 'error',
+    error: Object
+  }
```
Comment (meganle) [inline src/components/actions/LectureActions.js]: semicolons

File: src/components/common/constants.js [hunk 19]
```diff
@@ -1,10 +1,28 @@
...
+  MULTIPLE_CHOICE: 'MULTIPLE_CHOICE',
+  FREE_RESPONSE: 'FREE_RESPONSE',
+  MULTIPLE_ANSWER: 'MULTIPLE_ANSWER',
+  RANKING: 'RANKING'
+};
+
+export {
+  QUESTION_TYPES,
+  COLORS
+}
```
Comment (meganle) [inline src/components/common/constants.js]: semicolon

**Pull request merged**